From 29dc0928ee04c28a113327e120704d715d06c9c4 Mon Sep 17 00:00:00 2001
From: CMY <cmy@rock-chips.com>
Date: Tue, 10 Mar 2015 15:39:38 +0800
Subject: [PATCH] arm64: ion: fix scatterlist alloc fault on ion_system heap

Fix this bug:
WARNING: at lib/scatterlist.c:251 __sg_alloc_table+0x5c/0x15c()
Modules linked in: pvrsrvkm(O)
CPU: 2 PID: 582 Comm: Binder_5 Tainted: G           O 3.10.0 #10
Call trace:
[<ffffffc000088164>] dump_backtrace+0x0/0x140
[<ffffffc0000882b4>] show_stack+0x10/0x1c
[<ffffffc0008ccaf0>] dump_stack+0x1c/0x28
[<ffffffc0000a13c0>] warn_slowpath_common+0x74/0x9c
[<ffffffc0000a14ac>] warn_slowpath_null+0x14/0x20
[<ffffffc0002fadc4>] __sg_alloc_table+0x58/0x15c
[<ffffffc0002faee8>] sg_alloc_table+0x20/0x50
[<ffffffc00067ab7c>] ion_system_heap_allocate+0x188/0x318
[<ffffffc00067823c>] ion_alloc+0x134/0x788
[<ffffffc000678d68>] ion_ioctl+0x15c/0x344
[<ffffffc0001840ac>] vfs_ioctl+0x20/0x3c
[<ffffffc0001849b4>] do_vfs_ioctl+0x478/0x550
[<ffffffc000184ae8>] SyS_ioctl+0x5c/0x88

Signed-off-by: CMY <cmy@rock-chips.com>
---
 include/asm-generic/scatterlist.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/include/asm-generic/scatterlist.h b/include/asm-generic/scatterlist.h
index 5de07355fad4..8d3c787a946e 100644
--- a/include/asm-generic/scatterlist.h
+++ b/include/asm-generic/scatterlist.h
@@ -31,4 +31,8 @@ struct scatterlist {
 #define sg_dma_len(sg)		((sg)->length)
 #endif
 
+#ifdef CONFIG_ARCH_ROCKCHIP
+#define ARCH_HAS_SG_CHAIN
+#endif
+
 #endif /* __ASM_GENERIC_SCATTERLIST_H */
-- 
2.35.3

